### ISSUE TITLE
Sd2 returning NaN

### ISSUE THREAD

**Issue by ynjacobs** [rhenanbartels/hrv#27]
Hello and thanks so much for this wonderful library, it's been extremely helpful!

I'm trying to get `non_linear` feaures from a set of RRI values I extracted from the [Physionet Long Term AF Database](https://physionet.org/content/ltafdb/1.0.0/), but I am getting `sd2` returning `NaN` and I'm unsure as to why.

Here is the rri set (in seconds):
```
[0.57, 1.125, 0.57, 1.133, 0.57, 1.148, 0.57, 1.156, 0.586, 1.141, 0.562, 1.148, 0.562, 1.164, 0.57, 1.148, 0.57, 1.156, 0.562, 1.148, 0.57, 1.133, 0.578, 1.141, 0.578, 1.141, 0.562, 1.148, 0.57, 1.141, 0.57, 1.141, 0.57, 1.148, 0.562, 1.148, 0.562, 1.125, 0.562, 1.133, 0.57, 1.125, 0.578, 1.133, 0.562, 1.133, 0.578, 1.141, 0.57, 1.125, 0.57, 1.133, 0.57, 1.117, 0.57, 1.125, 0.578, 1.117, 0.578]
```
And here is how I am trying to get the Non Linear Features:
```
import numpy as np
from hrv.io import read_from_text
from hrv.rri import RRi
from hrv.classical import frequency_domain
from hrv.classical import non_linear
from hrv.classical import time_domain
import os

f = open("testNan.txt", "r")
lines = f.readlines()
averageArray = []
for l in lines[1:]:
    averageArray.append(l.split()[-3])
        #end for
averageArray = list(map(float, averageArray))

rri = RRi(averageArray)
      
nonLinear = non_linear(rri)
```
Wondering if `Nan` is expected as I'd like to use sd2 as a column in pandas dataframe but I get errors when having NaN values(and I wanna avoid dropping the whole row containing NaN)

Thanks so much!

**Comment by rhenanbartels** [comment]
Hi @ynjacobs, how are you doing?
Thank you very much for the kind words.

### Short example
if your RRi signal is the one in the array above, you can do the following:

```python
from hrv.classical import non_linear

rri = [
    0.57, 1.125, 0.57, 1.133, 0.57, 1.148, 0.57, 1.156, 0.586, 1.141, 0.562, 
    1.148, 0.562, 1.164, 0.57, 1.148, 0.57, 1.156, 0.562, 1.148, 0.57, 1.133, 
    0.578, 1.141, 0.578, 1.141, 0.562, 1.148, 0.57, 1.141, 0.57, 1.141, 0.57,
    1.148, 0.562, 1.148, 0.562, 1.125, 0.562, 1.133, 0.57, 1.125, 0.578, 1.133, 
    0.562, 1.133, 0.578, 1.141, 0.57, 1.125, 0.57, 1.133, 0.57, 1.117, 0.57, 1.125, 
    0.578, 1.117, 0.578
]

results = non_linear(rri)
print(results)

{'sd1': 0.39876285360774866, 'sd2': 0.05180883684177596}
```

The file `testNan.txt`  contains the same values as the rri array? If not, could you send me a sample file like `textNan.txt`?
It seems that the trick is here `averageArray.append(l.split()[-3])`

**Comment by ynjacobs** [comment]
Hi @rhenanbartels thanks so much for your reply!

Yes `testNan.txt` contains the same rri values as the rri array. It looks like:
```
1447748	A	1.125	N	1447892
1447892	N	0.570	A	1447965
1447965	A	1.125	N	1448109
1448109	N	0.570	A	1448182
1448182	A	1.133	N	1448327
1448327	N	0.570	A	1448400
1448400	A	1.148	N	1448547
1448547	N	0.570	A	1448620
1448620	A	1.156	N	1448768
1448768	N	0.586	A	1448843
1448843	A	1.141	N	1448989
1448989	N	0.562	A	1449061
1449061	A	1.148	N	1449208
1449208	N	0.562	A	1449280
1449280	A	1.164	N	1449429
1449429	N	0.570	A	1449502
1449502	A	1.148	N	1449649
1449649	N	0.570	A	1449722
1449722	A	1.156	N	1449870
1449870	N	0.562	A	1449942
1449942	A	1.148	N	1450089
1450089	N	0.570	A	1450162
1450162	A	1.133	N	1450307
1450307	N	0.578	A	1450381
1450381	A	1.141	N	1450527
1450527	N	0.578	A	1450601
1450601	A	1.141	N	1450747
1450747	N	0.562	A	1450819
1450819	A	1.148	N	1450966
1450966	N	0.570	A	1451039
1451039	A	1.141	N	1451185
1451185	N	0.570	A	1451258
1451258	A	1.141	N	1451404
1451404	N	0.570	A	1451477
1451477	A	1.148	N	1451624
1451624	N	0.562	A	1451696
1451696	A	1.148	N	1451843
1451843	N	0.562	A	1451915
1451915	A	1.125	N	1452059
1452059	N	0.562	A	1452131
1452131	A	1.133	N	1452276
1452276	N	0.570	A	1452349
1452349	A	1.125	N	1452493
1452493	N	0.578	A	1452567
1452567	A	1.133	N	1452712
1452712	N	0.562	A	1452784
1452784	A	1.133	N	1452929
1452929	N	0.578	A	1453003
1453003	A	1.141	N	1453149
1453149	N	0.570	A	1453222
1453222	A	1.125	N	1453366
1453366	N	0.570	A	1453439
1453439	A	1.133	N	1453584
1453584	N	0.570	A	1453657
1453657	A	1.117	N	1453800
1453800	N	0.570	A	1453873
1453873	A	1.125	N	1454017
1454017	N	0.578	A	1454091
```
So I use `averageArray.append(l.split()[-3])` to extract only the rri values to `averageArray[]`

**Comment by rhenanbartels** [comment]
Hi again @ynjacobs, sorry for the late reply.

I've checked the reason why SD2 is returning NaN and it is due to a negative value resulting from the expression:

```math
SD2 = sqrt(2 * std(rri) ** 2 - 0.5 * std(diff_rri) ** 2)
2 * std(rri) ** 2 - 0.5 * std(diff_rri) ** 2=  -103.9827432402817
```
And the square root of -103.9827432402817 is evaluating to a NaN value.

Regarding the `read_from_text` function,  I will fix this behavior. This looks like a bug.

Cheers,
R

**Comment by ynjacobs** [comment]
Thanks so much @rhenanbartels!

So is a negative value from the expression expected?

Thanks re `read_from_text`

Thanks again for all your help!

**Comment by rhenanbartels** [comment]
Unfortunately for this RRi series, yes.

This paper explains in a nice way how SD1 and SD2 are calculated:

> M. Brennan, M. Palaniswami, and P. Kamen. Do existing measures of Poincaré plot geometry reflect nonlinear features of heart rate variability. IEEE Trans Biomed Eng, 48(11):1342–1347, 2001.

Regarding the `read_from_text`, I've fixed the bug in version 0.2.10.

Cheers!
